services/alerta/client.go

```diff
@@ -0,0 +1,218 @@
+
+type Tags []string
+
+type Service []string
+
+type Alert struct {
+	Resource    string   `json:"resource"`    // warehouse-upload-aborted
+	Event       string   `json:"event"`       // CRITICAL/rudder/<resource>:<group>
+	Environment string   `json:"environment"` // PRODUCTION,DEVELOPMENT,PROXYMODE,CARBONCOPY
+	Severity    Severity `json:"severity"`    // warning,critical,normal // Get the full list from https://docs.alerta.io/api/alert.html#severity-table
+	Group       string   `json:"group"`       // cluster=rudder,destID=27CHciD6leAhurSyFAeN4dp14qZ,destType=RS,namespace=hosted,notificationServiceMode=CARBONCOPY,priority=P1,sendToNotificationService=true,team=warehouse
+	Text        string   `json:"text"`        // <event> is CRITICAL warehouse-upload-aborted-count: 1
+	Service     Service  `json:"service"`     // {upload_aborted}
+	Timeout     int      `json:"timeout"`     // 86400
+	Tags        Tags     `json:"tags"`        // {sendToNotificationService=true,notificationServiceMode=CARBONCOPY,team=warehouse,priority=P1,destID=27CHciD6leAhurSyFAeN4dp14qZ,destType=RS,namespace=hosted,cluster=rudder}
+}
+
+type Priority string
+
+type Severity string
+
+const (
+	SeverityCritical Severity = "critical"
+	SeverityWarning  Severity = "warning"
+	SeverityNormal   Severity = "normal"
+	SeverityOk       Severity = "ok"
+)
+
+const (
+	PriorityP1 Priority = "P1"
+	PriorityP2 Priority = "P2"
+	PriorityP3 Priority = "P3"
+)
+
+var (
+	defaultTimeout    = 30 * time.Second
+	defaultMaxRetries = 3
+	defaultSeverity   = SeverityCritical
+	defaultPriority   = PriorityP1
+	defaultTeam       = "No Team"
+)
+
+type Client interface {
+	SendAlert(ctx context.Context, resource string, service Service) error
+}
+
+type ClientImpl struct {
+	client   *http.Client
+	severity Severity
+	priority Priority
+	retries  int
+	url      string
+	text     string
+	tags     Tags
+	team     string
+}
+
+func WithHTTPClient(httpClient *http.Client) OptFn {
+	return func(c *ClientImpl) {
+		c.client = httpClient
+	}
+}
+
+func WithTimeout(timeout time.Duration) OptFn {
+	return func(c *ClientImpl) {
+		c.client.Timeout = timeout
+	}
+}
+
+func WithMaxRetries(retries int) OptFn {
+	return func(c *ClientImpl) {
+		c.retries = retries
+	}
+}
+
+func WithSeverity(severity Severity) OptFn {
+	return func(c *ClientImpl) {
+		c.severity = severity
+	}
+}
+
+func WithText(text string) OptFn {
+	return func(c *ClientImpl) {
+		c.text = text
+	}
+}
+
+func WithTags(tags Tags) OptFn {
+	return func(c *ClientImpl) {
+		c.tags = tags
+	}
+}
+
+func WithPriority(priority Priority) OptFn {
+	return func(c *ClientImpl) {
+		c.priority = priority
+	}
+}
+
+func WithTeam(team string) OptFn {
+	return func(c *ClientImpl) {
+		c.team = team
+	}
+}
+
+func NewClient(baseURL string, fns ...OptFn) *ClientImpl {
+	c := &ClientImpl{
+		url: baseURL,
+		client: &http.Client{
+			Timeout: defaultTimeout,
+		},
+
+		retries:  defaultMaxRetries,
+		severity: defaultSeverity,
+		priority: defaultPriority,
```

Comment:
severity and priority don't match with the client lifecycle. For example, should we create a new client if we want to send a different severity/priority event?

I would suggest removing them and moving them to `SendAlert` method.

```suggestion
```

Reply:
Great point.